### PR TITLE
Test separately built catalogues

### PR DESCRIPTION
**Pull request by bcumming**
* add a test that tests that a catalogue built separately can be loaded via the Python interface
* further simplification of dynamic library support
  * move all platform-specific code into the cpp implementation and out of header.

### CODE REVIEW

File: arbor/util/dylib.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <arbor/arbexcept.hpp>
+
+#include "util/dylib.hpp"
+#include "util/strprintf.hpp"
+
+namespace arb {
+namespace util {
+
+struct dl_handle {
```
Comment (thorstenhater) [inline arbor/util/dylib.cpp:14]: This is redundant now, it was meant as an opaque handle, but if it's not exposed, it can go.

File: scripts/test-catalogue.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+args = P.parse_args()
+catname = args.catname
+
+print(catname)
+
+if not os.path.isfile(catname):
+    print('ERROR: unable to open catalogue file')
+    sys.exit(1)
+
+print(ctypes.CDLL(catname))
```
Comment (thorstenhater) [inline scripts/test-catalogue.py:22]: This is also redundant: Debugging left-overs.

File: arbor/util/dylib.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    return {result};
+}
+
+void dl_close(dl_handle& handle) {
+    dlclose(handle.dl);
+    handle.dl = nullptr;
+}
+
+namespace impl{
+void* dl_get_symbol(const std::string& fn, const std::string& symbol) {
```
Comment (thorstenhater) [inline arbor/util/dylib.cpp:42]: What's the reason for the change from 
```cxx
void* dl_get_symbol(void* handel, const std::string& symbol)
```
? Now, fetching a symbol will go through `dlopen` everytime.

File: scripts/test-catalogue.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
+
+import argparse
+import ctypes
```
Comment (thorstenhater) [inline scripts/test-catalogue.py:4]: This can also go.

**Review by thorstenhater — CHANGES_REQUESTED** [review]

File: arbor/util/dylib.cpp (hunk 3 above)
Comment (bcumming) [reply arbor/util/dylib.cpp:42]: The library was only being used to return a symbol from a library, for which the `handle` type was not required.

This way the handle type is not required in the public interface. It could be reintroduced later, if we ever have more elaborate use cases for dynamic libraries.

**Comment by noraabiakar** [comment]
This PR updates pybind11, can we add that to the PR message with the version info?

**Comment by bcumming** [comment]
@noraabiakar  I can't understand why the PR shows pybind11 being updated: the pybind11 commit in the branch and in master are exactly the same, and the changes shown in the "Files changed" tab show pybind11 being updated to that same commit. 

So I don't think anything is actually being changed. It looks like GitHub is getting confused by the slightly convoluted history of the commits in this PR.

File: arbor/util/dylib.cpp (hunk 3 above)
Comment (thorstenhater) [reply arbor/util/dylib.cpp:33]: Good point.

File: arbor/util/dylib.hpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+// Find and return a symbol from a dynamic library with filename.
+// Throws dl_error on error.
+template<typename T>
+T dl_get_symbol(const std::string& filename, const std::string& symbol) {
+    return reinterpret_cast<T>(impl::dl_get_symbol(filename, symbol));
+}
+
+} // namespace util
+} // namespace arbo
```
Comment (thorstenhater) [inline arbor/util/dylib.hpp:26]: `arbo`?

**Review by thorstenhater — CHANGES_REQUESTED** [review]
Apart from a tiny typo and the pybind divergence everything's well.

**Review by thorstenhater — APPROVED** [review]
pybind11 malaise clarified offline, all good.
